File: applications/luci-statistics/luasrc/model/cbi/luci_statistics/irq.lua
--[[

Luci configuration model for statistics - collectd irq plugin configuration
(c) 2008 Freifunk Leipzig / Jo-Philipp Wich <xm@leipzig.freifunk.net>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

$Id$

]]--

m = Map("luci_statistics",
	translate("IRQ Plugin Configuration"),
	translate(
		"The irq plugin will monitor the rate of issues per second for " ..
		"each selected interrupt. If no interrupt is selected then all " ..
		"interrupts are monitored."
	))

-- collectd_irq config section
s = m:section( NamedSection, "collectd_irq", "luci_statistics" )

-- collectd_irq.enable
enable = s:option( Flag, "enable", translate("Enable this plugin") )
enable.default = 0

-- collectd_irq.irqs (Irq)
irqs = s:option( Value, "Irqs", translate("Monitor interrupts") )
irqs.optional = true
irqs:depends( "enable", 1 )

-- collectd_irq.ignoreselected (IgnoreSelected)
ignoreselected = s:option( Flag, "IgnoreSelected", translate("Monitor all except specified") )
ignoreselected.default  = 0
ignoreselected.optional = "true"
ignoreselected:depends( "enable", 1 )

return m

File: applications/luci-app-statistics/htdocs/luci-static/resources/view/statistics/plugins/irq.js
'use strict';
'require baseclass';
'require fs';
'require form';

return baseclass.extend({
	title: _('IRQ Plugin Configuration'),
	description: _('The irq plugin will monitor the rate of issues per second for each selected interrupt. If no interrupt is selected then all interrupts are monitored.'),

	addFormOptions: function(s) {
		var o;

		o = s.option(form.Flag, 'enable', _('Enable this plugin'));

		o = s.option(form.DynamicList, 'Irqs', _('Monitor interrupts'));
		o.optional = true;
		o.multiple = true;
		o.depends('enable', '1');
		o.load = function(section_id) {
			return fs.trimmed('/proc/interrupts').then(L.bind(function(str) {
				var lines = str.split(/\n/),
				    cpus = L.toArray(lines[0]);

				for (var i = 1; i < lines.length; i++) {
					var line = lines[i],
					    m = lines[i].match(/^\s*([^\s:]+):/);

					if (!m)
						continue;

					line = line.replace(/^[^:]+:\s+/, '');

					for (var j = 0; j < cpus.length; j++)
						line = line.replace(/^\d+\s*/, '');

					var desc = line.split(/ {2,}/).join(', ');

					this.value(m[1], '%s (%s)'.format(m[1], desc || '-'));
				}

				return this.super('load', [section_id]);
			}, this));
		};

		o = s.option(form.Flag, 'IgnoreSelected', _('Monitor all except specified'));
		o.depends('enable', '1');
	},

	configSummary: function(section) {
		var irqs = L.toArray(section.Irqs),
		    invert = section.IgnoreSelected == '1';

		if (irqs.length == 0)
			return _('Monitoring all interrupts');
		else if (invert)
			return N_(irqs.length, 'Monitoring all but one interrupt', 'Monitoring all but %d interrupts').format(irqs.length);
		else
			return N_(irqs.length, 'Monitoring one interrupt', 'Monitoring %d interrupts').format(irqs.length);
	}
});
